man/dhisexpire.8
.\" $Revision: 1.3 $
.TH DHISEXPIRE 8
.SH NAME
dhisexpire \- Diablo program to clean the history file
.PP
.SH SYNOPSIS
.B dhisexpire
[
.B \-a
[
.B \-m
]
[
.B \-o old-dhistory-file
]
[
.B \-p
]
[
.B \-r remember
]
[
.B \-T seconds
]
[
.B \-u
]
[
.B \-v
]
[
.B \-x
]
[
.B \-C diablo.config
]
[
.B \-d[n]
]
[
.B \-V
]
dhistory-file
[
new-dhistory-file
]

.SH DESCRIPTION
.IR DHisExpire 
This program is used to remove stale history entries by doing a
complete history rebuild. This process is required on a regular basis
(usually weekly) to prevent the history file from growing too big.
This program has been written to prevent any downtime to the server,
apart from a short pause at the end while the history file is
switched over (if the ``-a'' option is used). Incoming connections
are not closed, unless an error is detected.
.PP
Note that if the ``-a'' option is not used, the new history file
could become stale immediately and should not be used if the
diablo server is accepting incoming articles, otherwise some articles
will not be recorded in the new history.
.PP
When disabling the live history file to switch to the new one, the
history header magic flag is modified to prevent diablo processes from
using the file (they auto-detect and switch to the new one if
available). The ``-u'' option can be used on the old history file
to set the correct header magic, so that it can be used again.
.PP
The name of the history file is a required option. The new history
path/filename can be (optionally) specified as an extra option.
.PP
.B \-a
.PP
This option is required if the rebuilt history file is to be
used. The old history is renamed with ``.bak'' tagged onto the
and the history header magic flag is marked as invalid. By default,
the expire process doesn't move the new history file into place
and creates the new history file with ``.new'' tagged into the end.
.PP
.B \-m
.PP
By default, dhisexpire can only rename files on the same filesystem
Using this option, causes dhisexpire to do a full copy of the file
rather than a rename, allowing the file to be moved across filesystems.
.PP
.B \-o old-dhistory-file
.PP
Specify the path/filename to use to store the backup of the old history
file. By default, the file is path_to_history/dhistory.bak. Specifying
a filename of ``0'' causes dhisexpire to not make a backup of the
old history file.
.PP
.B \-p
.PP
With this option, the progress of the expire is displayed to stderr.
Note that if the diablo server is running and receiving articles,
the total entries may be lower than the actual entries and will be
adjusted when dhisexpire reaches the end.
.PP
.B \-r nn
.PP
Set the number of days that expired articles are kept in the history
file. The default is obtained from the ``rememberdays'' value in
diablo.config.
.PP
.B \-T nn
.PP
This option can be used to prevent old history entries from being
moved to the new history fule. The number specified is the number
of seconds that entries are valid and anything older will be
dropped from the history. WARNING: The old entries cannot be
recovered. This option is not normally used.
.PP
.B \-u
Set the history header magic flag to be valid again. When the history
rewrite is complete, the history header magic flag is marked as
invalid to prevent live processes from using it and to indicate to
them that a new history file is available.
.PP
.B \-v
This option turns on verbose mode.
.PP
.B -C
allows the specification of the path of the diablo.config file to use.
.PP
.B -d[n]
will set debug mode, at optional level n.
.PP
.B -V
will print the version and exit.
.PP

.SH "SEE ALSO"
diablo(8), 
dicmd(8),
didump(8),
diload(8),
dnewslink(8),
doutq(8),
dexpire(8),
dexpireover(8),
diconvhist(8),
dilookup(8),
dspoolout(8),
dkp(8),
dpath(8),
diablo-kp(5),
diablo-files(5)
.PP



RELEASE_NOTES

			    RELEASE NOTES

		  DIABLO NEWS TRANSIT AND READER SYSTEM

		      diablo-users@openusenet.org
		    http://www.openusenet.org/diablo/
			   news.software.nntp

Known Issues:
	- OLD: dreaderd: problem with corrupt header files caused
	  by what seems to be a lock race condition with 'dexpireover -R'.
	  This problem has existed for years.

Experimental (features/utilities that are new and/or not properly tested
  	      and hence should be used with care on production systems):
	- Compressed spools
	- Compressed feeds
	- IPv6 support
	- dnewsfeeds %include option
	- dexpirecache
	- dcancel

Summary of changes likely to impact an upgrade from 5.x
	- new dreader header index format. Cannot go back to
	  older versions as they will not be able to read
	  the new version.

Summary of changes likely to impact an upgrade from 4.x
	- history version 2 (offsets converted to indexes)
	- new history expire (dhisexpire)
	- fixed bug in dexpire.ctl inheritance handling

Summary of changes likely to impact an upgrade from 3.x
	- New spool storage format. Each article is stored with a binary
	  header. This new format is not usable with older versions of
	  diablo, although 4.x can read prior storage formats.

Diablo 5-current

2009-05-25
	* dreaderd: Don't know why no one ever noticed that tcp
	  socket options per user weren't getting set right.

2009-04-20
        * dcancel: Add the ability to cancel articles by article
          number in the reader header index. (from Michael Dorrington)

2009-04-18
	* Oops, never committed the stunnel patch for useproxied.

2009-02-01
	* parsedate wouldn't deal well with certain days of the
	  week that weren't capitalized

2009-01-28
	* Cleanup of various code problems, including new expire
	  cleanup, spaces-in-Newsgroups cleanup, etc.

2009-01-15
	* Implement alternate article numbering views for clients
	  (this version still needs some cleanup).  Clients may
	  indicate their capabilities via the XNUMBERING extension.

	  RFC977 imposes no real limits on article numbering.  This
	  may cause some issues with non-64-bit-clean clients.

	  RFC3977 imposes a 2^31-1 limit AND places restrictions
	  on the resetting of the low-water mark, which is excessively
	  retarded because it effectively means that once a group hits
	  the limit, it's just dead, but it's the behaviour that is
	  defined in the RFC.

	  WINDOW guarantees compatibility with non-64-bit-clean
	  clients by restricting the article numbers to a window
	  between 1 and 2^31-1; further, the number of articles in
	  a group is limited to guarantee that most clients should
	  detect it as a "group reset" along the various lines of
	  ill-defined logic often used by clients.  It is not a
	  frequent event, but when it happens, this has the best
	  possibility of being successful without being too
	  disruptive.

	  Another patch shortly will clean this code up, and will
	  allow for selection of which view to present by default
	  to clients who do not explicitly request a particular
	  view.

2008-12-24
	* Absolutely huge changes that have been underway for more
	  than half a year.  

	  There are a set of interrelated but different problems
	  going on in the system.

	  The first is that the legacy overview code was not 64-bit
	  safe; this effectively limited overviews for a group to
	  around 60 million articles and then wackiness happened.  
	  The code has been walked through and updated so that on 
	  64-bit platforms, larger overviews are possible (we've 
	  exercised it to 500 mil). 32-bit platforms currently lose
	  due to the inability of the operating system to map large 
	  files, but support for this is also possible and a cleanly 
	  coded solution may be committed in the future.

	  Further, article numbering is a disaster waiting to happen.
	  The legacy code is organized to handle 31-bit message 
	  numbers.  Certain alt.binaries high volume groups are 
	  poised to pass message number 2^31-1.  Now, on one hand,
	  the problem could be solved by simply 64-bit-ifying the
	  code to handle 63-bit article numbers.  On the other hand,
	  some nimrods decided to specify 2^31-1 as a maximum article
	  number in RFC3977, which, fortunately, we don't really 
	  claim compliance with.  This code revision simply extends
	  the article numbering to 2^63.  In doing so, it will update
	  your overview files from VERSION 3 to VERSION 4.  This is
	  a harmless change except for groups where article numbering
	  exceeds 2^31-1.  Some effort was put into keeping the files
	  otherwise-compatible with older code.

	  Note that if you are using Diablo as an article numbering
	  master, you will need to update Diablo AND dreaderd.

	  People running this codebase will be able to have larger
	  overviews and won't run into a problem at artno 2^31-1.

2008-12-22
	* Various compiler warning cleanups

2008-12-19
	* Runaway spool thread status flag repurposing, based on a 
	  strategy POC'd by tim@euroaccess (thanks for the headaches, 
	  grin)

2008-12-18
	* Various compiler warning cleanups

2008-12-18
	* It's 2008.  Everyone has zlib.  If I don't hear objections, I
	  will probably be slowly removing conditional support for it.

2008-12-18
	* Add error checking for group resizing; many platforms limit
	  calloc memory availability and application of an "unlimit" or
	  equivalent in the daily.reader script might be required.  This
	  just adds a reasonable error message

2008-12-18
	* Add compressed header transmission to clients (XZVER support)
	  Newshosting and a few other places have been running this for
	  almost a year without any complaints, but it is still in
	  testing.

2008-12-16
	* drgpctl masscancel: support cancellation of large numbers of
	  articles from very large groups (one-pass through the overview
	  file)

2008-12-15
	* Dump threads status from sigUsr1 (can't find who submitted this
	  patch, apologies, whoever)

2008-12-15
	* Report virtual server name in log lines


util/dhisexpire.c

/*
 * DHISEXPIRE.C	- Expire the history file with minimal downtime
 *
 * (c)Copyright 2002, Russell Vincent, All Rights Reserved.  Refer to
 *    the COPYRIGHT file in the base directory of this distribution 
 *    for specific rights granted.
 */

#include "defs.h"

#define	USE_DEADMAGIC

int VerboseOpt = 0;
int QuietOpt = 0;
int ShowProgress = 0;
int DonePause = 0;
int UseNewHistory = 0;
int DoRemember = 1;
int UnDead = 0;
int FileCopy = 0;
time_t MaxAge = -1;
int MustExit = 0;
char *FileName = NULL;
char NewFileName[PATH_MAX];
char OldFileName[PATH_MAX];
int HistoryVersion = 0;

void DoUnDead(int fd);
void DoExpire(int fd, int hsize, int rsize);
int ServerCmd(char *cmd);
void Fail(char *fname, char *errmsg);

void
Usage(void)
{
    printf("Expire old entries in the history file.\n\n");
    printf("dhisexpire [-a] [-p] [-r remember] [-T seconds] [-v] [-x]\n");
    printf("           [-C diablo.config] [-d[n]] [-V] dhistory-file [new-history]\n");
    printf("  where:\n");
    printf("\t-a\t- rename the new history to old history when finished\n");
    printf("\t-m\t- rename history files across filesystems (file copy)\n");
#ifndef	USE_DEADMAGIC
    printf("\t-P\t- don't pause diablo server\n");
#endif
    printf("\t-ofile\t- set path/name for backup of old history file\n");
    printf("\t-p\t- show progress on stdout\n");
    printf("\t-rN\t- set rememberdays\n");
    printf("\t-TN\t- don't dump articles older than N seconds\n");
    printf("\t-u\t- remove dead flag for a history file\n");
    printf("\t-v\t- be a little more verbose\n");
    printf("\t-x\t- keep records older than rememberdays old\n");
    printf("\t-Cfile\t- specify diablo.config to use\n");
    printf("\t-d[n]\t- set debug [with optional level]\n");
    printf("\t-V\t- print version and exit\n");
    exit(1);
}

void
sigInt(int sigNo)
{
    printf("Exit signal caught - exiting\n");
    ++MustExit;
    if (MustExit > 3)
	exit(1);
}

int
main(int ac, char **av)
{
    int fd;
    int i;
    int hsize = 1024 * 1024;
    int rsize = sizeof(History);
    struct stat st;

    NewFileName[0] = 0;
    OldFileName[0] = 0;

    LoadDiabloConfig(ac, av);

    for (i = 1; i < ac; ++i) {
	char *ptr = av[i];

	if (*ptr != '-') {
	    if (FileName) {
		if (NewFileName[0] != 0) {
		    fprintf(stderr, "unexpected argument\n");
		    exit(1);
		}
		strncpy(NewFileName, ptr, sizeof(NewFileName) - 1);
		NewFileName[sizeof(NewFileName) - 1] = '\0';
		continue;
	    }
	    FileName = ptr;
	    continue;
	}
	ptr += 2;
	switch(ptr[-1]) {
	case 'a':
	    UseNewHistory = 1; 
	    break;
	case 'm':
	    FileCopy = 1; 
	    break;
#ifndef	USE_DEADMAGIC
	case 'P':
	    DonePause = 1;
	    break;
#endif
	case 'o':
	    strncpy(OldFileName, ptr, sizeof(OldFileName) - 1);
	    OldFileName[sizeof(OldFileName) - 1] = '\0';
	    break;
	case 'p':
	    ShowProgress = 1;
	    break;
	case 'q':
	    QuietOpt = 1;
	    break;
	case 'r':
	    if (!*ptr)
		ptr = av[++i];
	    DOpts.RememberSecs = TimeSpec(ptr, "d");
	    if (DOpts.RememberSecs == -1)
		Usage();
	    break;
	case 'T':
	    MaxAge = btimetol(*ptr ? ptr : av[++i]);
	    break;
	case 'u':
	    UnDead = 1;
	    break;
	case 'v':
	    if (*ptr)
		VerboseOpt = strtol(ptr, NULL, 0);
	    else
		++VerboseOpt;
	    break;
	case 'x':
	    DoRemember = 0;
	    break;
	/* Common options */
	case 'C':		/* parsed by LoadDiabloConfig */
	    if (*ptr == 0)
		++i;
	    break;
	case 'd':
	    DebugOpt = 1;
	    if (*ptr)
		DebugOpt = strtol(ptr, NULL, 0);
	    break;
	case 'V':
	    PrintVersion();
	    break;
	default:
	    fprintf(stderr, "illegal option: %s\n", ptr - 2);
	    Usage();
	}
    }

    if (FileName == NULL) {
	Usage();
    }
    if (NewFileName[0] == 0)
	sprintf(NewFileName, "%s.new", FileName);
    if (OldFileName[0] == 0)
	sprintf(OldFileName, "%s.bak", FileName);
    if (strcmp(OldFileName, "0") == 0)
	OldFileName[0] = 0;

    if ((fd = open(FileName, O_RDWR)) >= 0 && fstat(fd, &st) == 0) {
	/*
	 * new style history file has a header
	 */

	HistHead hh;

	if (UnDead) {
	    DoUnDead(fd);
	    close(fd);
	    return(0);
	}
	if (read(fd, &hh, sizeof(hh)) != sizeof(hh)) {
	    perror("Corrupted history file");
	    exit(1);
	}
	if (hh.hmagic != HMAGIC) {
	    fprintf(stderr, "Corrupted history file - bad magic\n");
	    exit(1);
	}
	if (hh.version != HVERSION) {
	    fprintf(stderr, "WARNING! Version mismatch file V%d, expecting V%d\n", hh.version, HVERSION);
	    fprintf(stderr, "dump may be invalid\n");
	}
	rsize = hh.henSize;
	hsize = hh.hashSize;
	HistoryVersion = hh.version;

	lseek(fd, hh.headSize, 0);

	if (!QuietOpt)
	    printf("Hash table has %d entries, record size %d\n",
	    hsize,
	    rsize
	);

	DoExpire(fd, hsize, rsize);

	close(fd);
    } else {
	perror("History open failed");
	exit(1);
    }
    return(0);
}

void
DoUnDead(int fd)
{
    HistHead hh = { 0 };

    if (lseek(fd, 0, SEEK_SET) == -1) {
	perror("Unable to seek to pos 0 in history");
	return;
    }
    if (read(fd, &hh, sizeof(hh)) != sizeof(hh)) {
	perror("Unable to read history header");
	return;
    }
    if (hh.hmagic != HDEADMAGIC) {
	fprintf(stderr, "History file not marked as dead\n");
	return;
    }
    if (hh.version != HVERSION) {
	 fprintf(stderr, "ERROR! Version mismatch file V%d, expecting V%d\n", hh.version, HVERSION);
	return;
    }
    hh.hmagic = HMAGIC;
    if (lseek(fd, 0, SEEK_SET) == -1) {
	perror("Unable to seek to pos 0 in history");
	return;
    }
    if (write(fd, &hh, sizeof(hh)) != sizeof(hh)) {
	perror("Unable to write history header");
	return;
    }
    printf("History file %s no longer marked as dead\n", FileName);
}

void
DoExpire(int fd, int hsize, int rsize)
{
    char *hbuf;
    int hlen = rsize * 4096;
    int n;
    int rememberMins = DOpts.RememberSecs / 60;
    uint32 gmt = time(NULL) / 60;
    off_t seekpos = 0;
    uint32 totalentries = 0;
    uint32 okcount = 0;
    uint32 count = 0;
    uint32 failed = 0;
    uint32 ExpireDropCount = 0;
    uint32 ExpireKeepCount = 0;
    uint32 MaxAgeCount = 0;
    int r;
    History *h;
    int finished = 0;
    struct timeval tstart;
    struct timeval tend;
    double elapsed;

    hbuf = (char *)malloc(hlen);
    if (hbuf == NULL) {
	fprintf(stderr, "Unable to malloc %d bytes (%s)\n", hlen,
							strerror(errno));
	exit(1);
    }

    if (HistoryVersion > 1)
	seekpos = lseek(fd, hsize * sizeof(HistIndex) + rsize, 1);
    else
	seekpos = lseek(fd, hsize * sizeof(HistIndex), 1);

    {
	struct stat st;

	if (fstat(fd, &st) == -1) {
	    fprintf(stderr, "Unable to fstat history: %s\n", strerror(errno));
	    exit(1);
	}
	totalentries = (uint32)(((off_t)st.st_size - seekpos) / rsize);
	if (!QuietOpt)
	    printf("History entries start at offset %lld, %d records\n",
				(long long)seekpos, totalentries);
    }

    HistoryOpen(NewFileName, HGF_FAST|HGF_NOSEARCH|HGF_EXCHECK);

    gettimeofday(&tstart, NULL);

    while (!finished || MustExit) {
	int i;

	n = read(fd, hbuf, hlen) / rsize;

	if (n == 0) {
#ifdef USE_DEADMAGIC
	    if (DonePause > 0) {
		if (DonePause == 2) {
		    DonePause = 5;
		    HistoryClose();
		    if (UseNewHistory) {
			if (OldFileName[0])
			    remove(OldFileName);
			if (FileCopy) {
			    if (OldFileName[0] && MoveFile(FileName, OldFileName) == -1)
				Fail(NewFileName, "Unable to move history backup");
			    if (MoveFile(NewFileName, FileName) == -1)
				Fail(NewFileName, "Unable to move new history");
			} else {
			    if (OldFileName[0] && link(FileName, OldFileName) == -1)
				Fail(NewFileName, "Unable to link history backup");
			    if (rename(NewFileName, FileName) == -1)
				Fail(NewFileName, "Unable to rename new history");
			}
		    }
		}
		finished = 1;
	    } else {
		off_t seekpos = lseek(fd, 0, SEEK_CUR);
		HistHead hh = { 0 };
		if (lseek(fd, 0, SEEK_SET) == -1)
		    Fail(FileName, "Unable to seek to pos 0 in history");
		if (read(fd, &hh, sizeof(hh)) == -1)
		    Fail(FileName, "Unable to read history header");
		hh.hmagic = HDEADMAGIC;
		if (lseek(fd, 0, SEEK_SET) == -1)
		    Fail(FileName, "Unable to seek to pos 0 in history");
		if (write(fd, &hh, sizeof(hh)) == -1) {
		    perror("historyheadwrite");
		    Fail(FileName, "Unable to write history header");
		}
		if (lseek(fd, seekpos, SEEK_SET) == -1)
		    Fail(FileName, "Unable to seek in history");
		DonePause = 2;
	    }
#else
	    if (DonePause > 0) {
		if (DonePause == 2) {
		    rsignal(SIGINT, sigInt);
		    rsignal(SIGHUP, sigInt);
		    rsignal(SIGTERM, sigInt);
		    rsignal(SIGALRM, SIG_IGN);
		    DonePause = 3;
		    if (ServerCmd("pause") == 0)
			Fail(NewFileName, "Unable to pause diablo server");
		    DonePause = 4;
		    HistoryClose();
		    if (UseNewHistory) {
			if (OldFileName[0])
			    remove(OldFileName);
			if (FileCopy) {
			    if (OldFileName[0] && MoveFile(FileName, OldFileName) == -1)
				Fail(NewFileName, "Unable to move history backup");
			    if (MoveFile(NewFileName, FileName) == -1)
				Fail(NewFileName, "Unable to move new history");
			} else {
			    if (OldFileName[0] && link(FileName, OldFileName) == -1)
				Fail(NewFileName, "Unable to link history backup");
			    if (rename(NewFileName, FileName) == -1)
				Fail(NewFileName, "Unable to rename new history");
			}
		    }
		    if (ServerCmd("go") == 0)
			Fail(NewFileName, "Unable to resume diablo server");
		    DonePause = 5;
		}
		finished = 1;
	    } else {
		if (ServerCmd("flush") == 0)
		    Fail(NewFileName, "Unable to flush diablo server");
		if (ServerCmd("readonly") == 0)
		    Fail(NewFileName, "Unable to set readonly mode");
		if (ServerCmd("flush") == 0)
		    Fail(NewFileName, "Unable to flush diablo server");
		DonePause = 2;
		sleep(2);
	    }
#endif
	    continue;
	}

	gmt = time(NULL) / 60;

	for (i = 0; i < n; ++i) {
	    if (count++ > totalentries)
		totalentries = count;
	    if (ShowProgress && ((count % 32768) == 0)) {
		printf("%u/%u  (%d%%)   \r", count, totalentries,
				(int)((double)count * 100 / totalentries));
	    }

	    h = (History *)(hbuf + i * rsize);

	    /*
	     * If we specified a maximum age, then deal with it
	     */
	    if (MaxAge != -1) {
		int32 dgmt = (gmt - h->gmt) * 60;	/* Delta seconds */
		if (dgmt > MaxAge) {
		    MaxAgeCount++;
		    continue;
		}
	    }

	    if (DoRemember && H_EXPIRED(h->exp)) {
		int32 dgmt = gmt - h->gmt;	/* DELTA MINUTES */
		if (dgmt < -rememberMins || dgmt > rememberMins) {
		    ExpireDropCount++;
		    continue;
		} else {
		    ExpireKeepCount++;
		}
	    }

	    if ((r = HistoryAdd(NULL, h)) == 0)
		++okcount;
	    else if (r != RCTRYAGAIN)
		++failed;
	    else
		Fail(NewFileName, "HistoryAdd: write failed!");
	}
    }

    gettimeofday(&tend, NULL);

    if (DonePause == 3 && ServerCmd("go") == 0)		/* Got signal */
	Fail(NewFileName, "Unable to resume diablo server");
    if (ShowProgress && totalentries > 0)
	printf("%u/%u  (%d%%)   \n", count, totalentries,
				(int)((double)count * 100 / totalentries));
    if (!QuietOpt) {
	elapsed = (tend.tv_sec + tend.tv_usec / 1000000.0) -
				(tstart.tv_sec + tstart.tv_usec / 1000000.0);
	if (elapsed == 0)
	    elapsed = 1;
        printf("%.3f seconds\n", elapsed);
        printf("%u entries processed (%.0f per second)\n",
					count, (double)count / elapsed);
        printf("%u entries kept\n", okcount);
        printf("%u entries failed\n", failed);
        printf("%u expired entries kept\n", ExpireKeepCount);
        printf("%u expired entries dropped\n", ExpireDropCount);
        printf("%u dropped as beyond max age\n", MaxAgeCount);
    }
}

int
ServerCmd(char *cmd)
{
    FILE *fi;
    FILE *fo;
    char buf[256];
    int r = 0;

    /*
     * UNIX domain socket
     */

    {
	struct sockaddr_un soun;
	int ufd;

	memset(&soun, 0, sizeof(soun));

	if ((ufd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
	    perror("udom-socket");
	    return(r);
	}
	soun.sun_family = AF_UNIX;
	sprintf(soun.sun_path, "%s", PatRunExpand(DiabloSocketPat));
	if (connect(ufd, (struct sockaddr *)&soun, offsetof(struct sockaddr_un, sun_path[strlen(soun.sun_path)+1])) < 0) {
	    perror("udom-connect");
	    return(r);
	}
	fo = fdopen(dup(ufd), "w");
	fi = fdopen(ufd, "r");
    }

    fprintf(fo, "%s\n", cmd);
    fprintf(fo, "quit\n");
    fflush(fo);
    while (fgets(buf, sizeof(buf), fi) != NULL) {
	if (VerboseOpt)
	    printf("%s", buf);
	if (strncmp(buf, "200", 3) == 0)
	    r = 1;
	if (strncmp(buf, "211 Flushing feeds", 18) == 0)
	    r = 1;
	if (strcmp(buf, ".\n") == 0)
	    break;
    }
    fclose(fo);
    fclose(fi);
    return(r);
}

void
Fail(char *fname, char *errmsg)
{
    printf("%s\n", errmsg);
    printf("History rebuild is not complete - keeping old history\n");
    HistoryClose();
    if (fname != NULL)
	remove(fname);
    exit(1);
}

